### ISSUE TITLE
Add terminfo class

### ISSUE THREAD

**Issue by johnbradley** [phenoscape/rphenoscape#192]
Add a terminfo class similar to the [phenotype class](https://github.com/phenoscape/rphenoscape/blob/master/R/phen_obj.R). 
The class should retrieve data from the kb.phenoscape API [/term endpoint](https://kb.phenoscape.org/apidocs/#/Terms/get_term) for an IRI.
If the /term response's label field is empty the class should fetch [/term/label endpoint](https://kb.phenoscape.org/apidocs/#/Terms/get_term_label) to populate this data.
Add the ability to optionally fetch data from [/term/classification endpoint](https://kb.phenoscape.org/apidocs/#/Terms/get_term_classification) using a boolean `withClassification` parameter.

**Comment by johnbradley** [comment]
For this my initial plan is to follow the pattern of the phenotype class.
The `as.terminfo` function takes either a IRI or a dataframe with IRI in the `id` column and returns a list of the result. I added a rough print method to show whatever data we have.

## Examples

Example using an IRI:
```
> as.terminfo('http://purl.obolibrary.org/obo/UBERON_0011618')
terminfo 'basihyal bone'
Definition: Replacement bone that is median and is the anterior-most bone of the ventral hyoid arch.
Synonyms:
    basihyoid
    glossohyal
Relationships:
    develops from basihyal cartilage
```
Example including optional classification data:
```
> as.terminfo('http://purl.obolibrary.org/obo/UBERON_0011618', withClassification = TRUE)
terminfo 'basihyal bone'
Definition: Replacement bone that is median and is the anterior-most bone of the ventral hyoid arch.
Synonyms:
    basihyoid
    glossohyal
Relationships:
    develops from basihyal cartilage
Subclass of:
    basihyal element
    endochondral bone
Superclass of:
    basihyal bone (zebrafish)
```

Example dataframe version using find_term:
```
> as.terminfo(find_term('basihyal bone'))
[[1]]
terminfo 'basihyal bone'
Definition: Replacement bone that is median and is the anterior-most bone of the ventral hyoid arch.
Synonyms:
    basihyoid
    glossohyal
Relationships:
    develops from basihyal cartilage

[[2]]
terminfo 'basihyal bone (zebrafish)'
...
```

## Code
I have a [192-adds-terminfo-class](https://github.com/phenoscape/rphenoscape/tree/192-adds-terminfo-class) branch with my code so far. [View as.terminfo in that branch.](
https://github.com/phenoscape/rphenoscape/blob/9cb20d364407281b25105b03b3a40f4f364d381b/R/terms.R#L176-L226)

**Comment by hlapp** [comment]
Nice! I think we should print the IRI too for each terminfo class. Could you also show the actual structure of the list that is of class `terminfo` to show names and their values.

I wanted to make some inline comments. Should I wait until you have a PR posted, or use the (or one of the) commit(s)?

**Comment by hlapp** [comment]
@wdahdul please feel free review and comment on the print format. This relates (and will ultimately be the solution) to #77 (it does not matter to R at all which element in a list a certain name, i.e., key is, so what you are asking about in #77 is a friendlier printout, which we're discussing here). Also see https://github.com/phenoscape/rphenoscape/issues/52#issuecomment-812614732.

**Comment by johnbradley** [comment]
@hlapp I just created a draft PR. You can comment inline there: https://github.com/phenoscape/rphenoscape/pull/195

**Comment by johnbradley** [comment]
Here is the structure of the list from the above examples.
Below is a screenshot from RStudio after running:
```
basihyal_bone <- as.terminfo('http://purl.obolibrary.org/obo/UBERON_0011618', withClassification = TRUE)
View(basihyal_bone)
```
![Screen Shot 2021-04-02 at 2 03 51 PM](https://user-images.githubusercontent.com/1024463/113441487-660a0700-93bc-11eb-9d34-282fff486913.png)

If withClassification = FALSE (the default) the entire `classification` section will be missing.

**Comment by hlapp** [comment]
Looks good so far. I'm wondering whether we shouldn't remove the redundant `id` and `label` from the `classification` element. I fear the moment when they are not exactly the same (nearly impossible for the `id`, but maybe not impossible for the `label`.

**Comment by wdahdul** [comment]
This is looking good! I agree with Hilmar about including IRI in the print output. 

For the relationship/classification info, I’m used to seeing these combined (for example how [OLS](https://www.ebi.ac.uk/ols/ontologies/uberon/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FUBERON_0011618) displays it). I see from the above examples that classification is as an optional output, but I think it's more useful to see both relationships and classification as the default, under the same Subclass of category, like below:
 
```
Subclass of: 
develops_from basihyal cartilage
basihyal element
endochondral bone

Superclass of:
basihyal bone (zebrafish)
```

**Comment by hlapp** [comment]
Thanks for the feedback @wdahdul! The reason for classification being optional is that it involves a separate API call for each term. This may not matter at all for one term IRI in interactive mode, but when doing this for 1000 in batch mode where the classification info isn't needed this will at least double the execution time, for no benefit to the user. Perhaps you are arguing that the default should be on, with the option to turn it off. It's normally good practice that default behavior is well-performing behavior. That said, perhaps the best middle ground here is to make it controllable over `options()`. (Thoughts on this, @johnbradley ?)

As for the print display, I'm hesitant to make this too smart, thereby not just pretty-printing but obscuring the way the data are structured internally. A `Relationships:` section would still convey the information in well-readable form, wouldn't it?

**Comment by wdahdul** [comment]
Sounds fine to me - the above are good arguments for keeping the classification as optional in the default, and same for the second point about displaying a Relationships section. I was thinking from the perspective of a UI user, but here an R user will be savvier about selecting different options.

**Comment by johnbradley** [comment]
@hlapp Using `options()` sounds like a good solution. I'll add a change to the branch for this.
Off topic: Should use `options()` for the [phenoscape_api URL](https://github.com/phenoscape/rphenoscape/blob/6b59eda736431a8f25e08969eebe2dcbbf3c3150/R/pkb_get.R#L215-L223) setting as well?

**Comment by hlapp** [comment]
> Should we use `options()` for the [phenoscape_api URL](https://github.com/phenoscape/rphenoscape/blob/6b59eda736431a8f25e08969eebe2dcbbf3c3150/R/pkb_get.R#L215-L223) setting as well?

I thought about this when working on #188. Ultimately I wasn't sure about the benefit on top of an environment variable, and fi we did implement support for an `options()` setting, then I think we would also need to change this so that the setting wouldn't persist for the remainder of the R session, but instead would be re-evaluated every single time. Possible, but again, I felt I would want to see a supporting use-case first where the environment variable doesn't cut it.

**Comment by johnbradley** [comment]
I'm not clear on how we should print the relationships. Should I just indent the Subclass of/Superclass of data?
For example:
```
Relationships:
    develops from basihyal cartilage
    Subclass of:
        basihyal element
        endochondral bone
    Superclass of:
        basihyal bone (zebrafish)
```

**Comment by hlapp** [comment]
Not indent. I think just as originally above is fine actually, and would maintain a closer resemblance of the actual nested data structure.

**Comment by johnbradley** [comment]
I was looking into how terminfo might work with a taxon. The `/term` endpoint does return data for a taxon IRI.
Example using `as.terminfo` with a taxon IRI:
```
> as.terminfo('http://purl.obolibrary.org/obo/VTO_9020379', withClassification = TRUE)
terminfo 'Diplocaulus' http://purl.obolibrary.org/obo/VTO_9020379
Synonyms:
    Permoplatyops (related)
Subclass of:
    Keraterpetontidae
Superclass of:
    Diplocaulus brevirostris
    Diplocaulus magnicornis
    Diplocaulus minimus
    Diplocaulus recurvatus
    Diplocaulus salamandroides
```
The pre-existing [taxon_info() function](https://github.com/phenoscape/rphenoscape/blob/3361811ede8ddbb2108e9666be60e0b60b77dbfb/R/terms.R#L34-L43) uses the `/taxon` endpoint returning `id`, `label`, `extinct`, and `rank`.
`terminfo` could fetch the `extinct` and `rank` data from the `/taxon` endpoint when the IRI is a taxon.
Updated example:
```
> as.terminfo(find_term('Coralliozetus', matchTypes=c('exact')), withClassification = TRUE)
terminfo 'Coralliozetus' http://purl.obolibrary.org/obo/VTO_0042955
Subclass of:
    Chaenopsidae
Superclass of:
    ...
    Coralliozetus springeri
Extinct: FALSE
Rank: genus
```
I think an IRI is a taxon if `isDefinedBy` == "http://purl.obolibrary.org/obo/vto.owl".

**Comment by hlapp** [comment]
> I think an IRI is a taxon if `isDefinedBy` == "http://purl.obolibrary.org/obo/vto.owl".

In first approximation yes. For a more general treatment, you can use `obo_ont_type()` (file `term-type.R`).

**Comment by johnbradley** [comment]
Here is the structure of the list with taxon data:
![Screen Shot 2021-04-09 at 9 14 45 AM](https://user-images.githubusercontent.com/1024463/114186457-e1b40880-9914-11eb-89b7-e98e03a0dbad.png)

**Comment by johnbradley** [comment]
I added common_name to the terminfo list data for taxon data after seeing this field is sometimes populated.
Updated print function with example:
```
as.terminfo('http://purl.obolibrary.org/obo/VTO_0042957')
terminfo 'Coralliozetus angelicus' http://purl.obolibrary.org/obo/VTO_0042957
Synonyms:
    Angel blenny (related)
    Emblemaria angelica (related)
Extinct: FALSE
Rank: species
Common Name: Angel blenny
```

**Comment by hlapp** [comment]
I'm wondering whether the `extinct` and `rank` keys shouldn't sit directly at the first level of `terminfo` nesting, i.e., be un-nested. They properties can't repeat (there's only one rank, and one "extinct", and they do represent direct properties of the "term" when the term represents a taxon.

**Comment by johnbradley** [comment]
I was thinking about creating a `as.data.frame()` method for the `terminfo` class.
Then you could get functionality similar to `taxon_info`, `anatomy_term_info`, and `phenotypic_quality_term_info`.
```
> terminfo <- as.terminfo(find_term("Coralliozetus angelicus"))
> as.data.frame(terminfo)
                    label                            isDefinedBy definition                                         id extinct
1 Coralliozetus angelicus http://purl.obolibrary.org/obo/vto.owl            http://purl.obolibrary.org/obo/VTO_0042957   FALSE
   common_name                                        rank.id rank.label
1 Angel blenny http://purl.obolibrary.org/obo/TAXRANK_0000006    species
```

**Comment by hlapp** [comment]
This would really only be useful if you could apply this to a list of terminfo objects, resulting in a data.frame with multiple rows, which would offer advantages for subsetting. If there's only a single row then there's subsetting and I don't think easier access to "columns" (values of keys in the list case) either.

I'm not sure though how you would use the `as.data.frame()` generic for list to achieve this. So maybe just a utility function?

**Comment by johnbradley** [comment]
> I'm not sure though how you would use the as.data.frame() generic for list to achieve this. So maybe just a utility function?

I was thinking `as.data.frame.terminfo` would apply, but you are correct it would have to be a utility function.

**Comment by johnbradley** [comment]
This was implemented in PR #195
